sim/message_queue_test.go

```diff
@@ -0,0 +1,158 @@
+	require.Equal(t, want2nd, subject.Remove())
+	require.Equal(t, want3rd, subject.Remove())
+	require.Equal(t, want4th, subject.Remove())
+	require.Zero(t, subject.Len())
+}
+
+func TestMessageQueue_ReturnsNilWhenEmpty(t *testing.T) {
+	subject := newMessagePriorityQueue()
+	wantMsg := &messageInFlight{
+		source:    0,
+		dest:      2,
+		payload:   "fish",
+		deliverAt: time.Time{}.Add(17 * time.Second),
+	}
+
+	subject.Insert(wantMsg)
+
+	require.Equal(t, subject.Len(), 1)
+	require.Equal(t, wantMsg, subject.Remove())
+	require.Equal(t, subject.Len(), 0)
+	require.Nil(t, subject.Remove())
+}
+
+func TestMessageQueue_ContainsAllDuplicateMessages(t *testing.T) {
+	const insertions = 5
+	subject := newMessagePriorityQueue()
+	msg := &messageInFlight{deliverAt: time.Time{}.Add(17 * time.Second)}
+	for i := 0; i < insertions; i++ {
+		subject.Insert(msg)
+	}
+	require.Equal(t, subject.Len(), insertions)
+	for i := 0; i < insertions; i++ {
+		require.Equal(t, msg, subject.Remove())
+	}
+	require.Equal(t, subject.Len(), 0)
+}
+
+func TestMessageQueue_UpsertFirstWhere(t *testing.T) {
+	// Sample values are ordered by their name, i.e. m1 is expected to be queued
+	// before m2, m2 before m3 and so on. This is to make reading the test cases and
+	// expected behaviour easier.
+	var (
+		m1 = &messageInFlight{
+			source:    17,
+			dest:      2,
+			payload:   "fish",
+			deliverAt: time.Time{}.Add(3 * time.Millisecond),
+		}
+		m2 = &messageInFlight{
+			source:    6,
+			dest:      7,
+			payload:   "lobster",
+			deliverAt: time.Time{}.Add(7 * time.Second),
+		}
+		m3 = &messageInFlight{
+			source:    88,
+			dest:      88,
+			payload:   "barreleye",
+			deliverAt: time.Time{}.Add(22 * time.Minute),
+		}
+		m4 = &messageInFlight{
+			source:    8989,
+			dest:      0,
+			payload:   "fishmonger",
+			deliverAt: time.Time{}.Add(63 * time.Hour),
+		}
+	)
+
+	tests := []struct {
+		name            string
+		givenInsertions []*messageInFlight
+		givenMatcher    func(*messageInFlight) bool
+		givenUpsert     *messageInFlight
+		wantMessages    []*messageInFlight
+	}{
+		{
+			name:            "matching none inserts",
+			givenInsertions: []*messageInFlight{m2, m4},
+			givenMatcher:    func(*messageInFlight) bool { return false },
+			givenUpsert:     m1,
+			wantMessages:    []*messageInFlight{m1, m2, m4},
+		},
+		{
+			name:            "matching one updates",
+			givenInsertions: []*messageInFlight{m1, m2, m4},
+			givenMatcher:    func(msg *messageInFlight) bool { return msg.payload == m1.payload },
+			givenUpsert:     m3,
+			wantMessages:    []*messageInFlight{m2, m3, m4},
+		},
+		{
+			name:            "matching more than one updates first",
+			givenInsertions: []*messageInFlight{m1, m2, m4},
+			givenMatcher:    func(msg *messageInFlight) bool { return msg.payload == m2.payload || msg.payload == m4.payload },
+			givenUpsert:     m3,
+			wantMessages:    []*messageInFlight{m1, m3, m4},
+		},
+	}
+	for _, test := range tests {
+		test := test
```

Comment:
Do we require Go 1.22? If so this is no longer required.

Reply:
Ok, no looks like we are on Go 1.21

Reply:
yep; can't wait.